### PR TITLE
Smooth intersection turning

### PR DESCRIPTION
**Pull request by marc7s**
* Made cars turn in intersections with support for occupancy
* Added DebugUtility
* Added placeholders for speed limit signs

### CODE REVIEW

File: TrafficSimulator/Assets/Prefabs/Car/AutoDrive.cs [hunk 1]
```diff
@@ -69,27 +71,32 @@ public class AutoDrive : MonoBehaviour
         private Vehicle _vehicle;
         private float _vehicleLength;
         private Bounds _vehicleBounds;
+        
+        private float _brakeDistance = 0;
+        
+        private float _originalMaxSpeed;
```
Comment (belgien) [inline TrafficSimulator/Assets/Prefabs/Car/AutoDrive.cs:77]: Maybe change to _startMaxSpeed

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 2]
```diff
@@ -40,17 +41,29 @@ public class Intersection : MonoBehaviour
...
+        [HideInInspector] private Dictionary<string, RoadNode> _intersectionGuideRoadNodes = new Dictionary<string, RoadNode>();
+        [HideInInspector] private Dictionary<string, LaneNode> _intersectionEntryNodes = new Dictionary<string, LaneNode>();
+        [HideInInspector] private Dictionary<string, LaneNode> _intersectionExitNodes = new Dictionary<string, LaneNode>();
+        [HideInInspector] private Dictionary<(string, string), GuideNode> _intersectionGuidePaths = new Dictionary<(string, string), GuideNode>();
         [HideInInspector] public IntersectionType Type;
         [ReadOnly] public string ID;
 
+        [Header("Connections")]
+        public GameObject GuideRoadNodePrefab;
+        public GameObject GuideLaneNodePrefab;
```
Comment (hannes44) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:54]: These two should be private

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 3]
```diff
@@ -569,45 +655,206 @@ public void MapIntersectionNavigation()
...
+
+        private RoadNode FetchOrGenerateRoadNodes(RoadNode start, RoadNode end)
+        {
+            // Check if this road node has already been generated
+            if(_intersectionGuideRoadNodes.ContainsKey(start.ID))
+                return _intersectionGuideRoadNodes[start.ID];
+            
+            const float roadNodeDistance = 3f;
+            
+            RoadNode head = CreateEvenlySpacedGuideRoadNodes(start.Position, end.Position, roadNodeDistance);
```
Comment (hannes44) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:758]:
```suggestion
            RoadNode head = CreateEvenlySpacedGuideRoadNodes(start.Position, end.Position, roadNodeDistance) ?? start;
```

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 4]
```diff
@@ -569,45 +655,206 @@ public void MapIntersectionNavigation()
...
+                Vector3 normal = Quaternion.Euler(0, 90, 0) * tangent;
+                float distanceToPrev = prev == null ? 0 : Vector3.Distance(prev.Position, position);
+                
+                curr = curr == null ? new RoadNode(position, tangent, normal, RoadNodeType.IntersectionGuide, 0, 0) : new RoadNode(position, tangent, normal, RoadNodeType.IntersectionGuide, prev, null, 0, distanceToPrev);
+                
+                if(prev != null)
+                    prev.Next = curr;
+                prev = curr;
+            }
+            return curr == null ? null : curr.First;
```
Comment (hannes44) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:788]:
```suggestion
           return curr?.First;
```

File: TrafficSimulator/Assets/Prefabs/Car/AutoDrive.cs [hunk 5]
```diff
@@ -539,6 +600,14 @@ private void P_UpdateTargetAndCurrent()
             UpdateTargetFromNavigation();
         }
 
+        private bool P_HasReachedTarget()
```
Comment (belgien) [inline TrafficSimulator/Assets/Prefabs/Car/AutoDrive.cs:603]: Function could be a single return

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 6]
```diff
@@ -549,13 +565,83 @@ private void AssignMeshComponents()
...
+                    int i = 0;
+                    while(curr != null && curr.Type == RoadNodeType.IntersectionGuide)
+                    {
+                        GameObject roadNodeObject = Instantiate(GuideRoadNodePrefab, curr.Position, curr.Rotation, _guideNodeContainer.transform);
+                        roadNodeObject.name = i + " " + curr.Type;
+
+                        curr = curr.Next;
+                        i++;
+                    }
+                }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs]: Similar code written twice

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 7]
```diff
@@ -569,45 +655,206 @@ public void MapIntersectionNavigation()
...
+            // Precompute all the guide paths and store them
+            foreach(LaneNode entry in entryNodes)
+            {
+                foreach(LaneNode exit in exitNodes)
+                {
+                    if(_intersectionEntryNodes.ContainsKey(entry.ID) && _intersectionExitNodes.ContainsKey(exit.ID))
+                        _intersectionGuidePaths.Add((entry.ID, exit.ID), GetGuidePath(entry, exit));
+                }
+            }
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:705]:
```suggestion
        }
        
```

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Utility/DebugUtility.cs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            _endPointPrefab.transform.position = farAway;
+            _endPointPrefab.GetComponent<Renderer>().material = new Material(Shader.Find("Standard"));
+
+            _lineContainer = new GameObject(_lineContainerName);
+            _lineRenderer = _lineContainer.AddComponent<LineRenderer>();
+            _lineRenderer.positionCount = 0;
+            _lineContainer.transform.parent = _container.transform;
+
+            _isSetup = true;
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Utility/DebugUtility.cs:57]: Needs comments

File: TrafficSimulator/Assets/Scenes/Examples/DebugUtility/DebugUtilityExample.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+
+namespace Examples
+{
+    public class DebugUtilityExample : MonoBehaviour
+    {
+        private Dictionary<string, (Vector3[], Quaternion[])> _positionGroups = new Dictionary<string, (Vector3[], Quaternion[])>();
+        private Vector3[] _linePositions = new Vector3[]{ Vector3.zero, new Vector3(10, 5, 10), new Vector3(15, 0, 10) };
+
+
```
Comment (belgien) [inline TrafficSimulator/Assets/Scenes/Examples/DebugUtility/DebugUtilityExample.cs:12]:
```suggestion
```

**Review by belgien — CHANGES_REQUESTED** [review]

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs (hunk 6 above)
Comment (marc7s) [reply TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs]: Would require a templated function which seems overkill

**Review by belgien — APPROVED** [review]

**Pull request merged**